### PR TITLE
Fixes issues where controller and server can clobber each other

### PR DESCRIPTION
**Pull request by alexbrand**
As described in #962, the controller and server overwrite their changes during operation. This is a problem because any change requested by the user can be wiped by the controller when it finishes an in-flight operation.

To avoid this issue, the controller should only be able to touch the current state field, while the server should only touch the desired state field. Given that these fields live in the same type and that controller operations are long-running, the controller needs to re-read the cluster object before updating it to pick up any changes that might have occurred while performing the operation. Ideally, this would be done in a transaction, but the current implementation of the store does not expose them. We will have to address this in a follow up issue.

Another outstanding issue is that of changing the plan file, as both the controller and server are currently manipulating the plan. We can get away with it for now because they touch different fields, but it might become a problem in the future. I am starting to think that the plan file should be considered an implementation detail that is handled by the controller, but this needs more thought.

### CODE REVIEW

**Comment by alexbrand** [comment]
After an initial iteration on this, I realized that I could decouple server and cluster even more by actually having them update different structs. That is what I am proposing here.

The server will write/update the cluster specification, whereas the controller will write/update the status.

**Comment by emmetthitz** [comment]
Makes sense to me. Is this still WIP then?

**Comment by alexbrand** [comment]
Yes, I think this is ready. 

There is some outstanding work that I want to complete in follow up PRs to not make this one even larger.

File: pkg/controller/cluster.go [hunk 1]
```diff
@@ -62,161 +108,248 @@ func (c *clusterController) run(watch <-chan struct{}) {
...
+		EtcdNodes:     clusterSpec.EtcdCount,
+		MasterNodes:   clusterSpec.MasterCount,
+		WorkerNodes:   clusterSpec.WorkerCount,
+		IngressNodes:  clusterSpec.IngressCount,
+	}
+	planner := &install.BytesPlanner{}
+	if err := install.WritePlanTemplate(planTemplate, planner); err != nil {
+		return fmt.Errorf("could not write plan template: %v", err)
+	}
+	var p *install.Plan
```
Comment (emmetthitz) [inline pkg/controller/cluster.go]: I think the next line makes this unnecessary?

File: pkg/controller/cluster.go [hunk 2]
```diff
@@ -62,161 +108,248 @@ func (c *clusterController) run(watch <-chan struct{}) {
...
+	// Set values in the plan
+	p.Cluster.Name = clusterName
+	p.Provisioner = install.Provisioner{Provider: clusterSpec.Provisioner.Provider}
+	// TODO: Handle provisioner specific options (e.g. AWS Region)
+	switch clusterSpec.Provisioner.Provider {
+	case "aws":
+		p.Provisioner.AWSOptions = &install.AWSProvisionerOptions{}
+		// nothing
+	case "azure":
+		p.AddOns.CNI.Provider = "weave"
```
Comment (emmetthitz) [inline pkg/controller/cluster.go]: Do we have to use weave on azure?

Comment (alexbrand) [reply pkg/controller/cluster.go]: We do, yes. Calico is not supported on Azure.

File: pkg/controller/cluster.go (hunk 1 above)
Comment (alexbrand) [reply pkg/controller/cluster.go]: Fixed. Thanks!

**Comment by dkoshkin** [comment]
LGTM!

**Pull request merged**